src/device/device_ledger.cpp

```diff
@@ -320,7 +320,9 @@ namespace hw {
     bool device_ledger::reset() {
       reset_buffer();
       int offset = set_command_header_noopt(INS_RESET);
-      memmove(this->buffer_send+offset, MONERO_VERSION, strlen(MONERO_VERSION));
+      const size_t verlen = strlen(MONERO_VERSION);
+      ASSERT_X(offset + verlen < BUFFER_SEND_SIZE, "MONERO_VERSION is too long")
+      memmove(this->buffer_send+offset, MONERO_VERSION, verlen);
       offset += strlen(MONERO_VERSION);
```

Comment:
```suggestion
       offset += verlen;
```

Reply:
If MONERO_VERSION is a string literal then you shouldn't be using a runtime computation (strlen) at all. You should be using sizeof() instead, since this is a compile-time constant.

In OpenLDAP we use `#define STRLENOF(X) sizeof(x)-1` (which accounts for the trailing NUL character)...

Reply:
The compiler is usually capable of recognizing string literals and converting `std::strlen` into constants. Recent versions of Clang and Gcc do this anyway. There may be situations where they fail to detect this (tested only on [godbolt.org](https://godbolt.org/)).

Reply:
That's wonderful, but relying on compiler magic instead of having explicitly written behavior is how you get unexpected holes in security-sensitive code.

Reply:
While we are discussing this code chunk, current Monero Ledger app has a restriction on version string length, 10 chars max (is not null terminated).
Might be worth it to add a static assert on max MONERO_VERSION length.

Check in Monero Ledger app code:
```c++
  char client_version[10];
  ...
  if (client_version_len > 10) {
    THROW(SW_CLIENT_NOT_SUPPORTED+1);
  }
```

Reply:
> That's wonderful, but relying on compiler magic instead of having explicitly written behavior is how you get unexpected holes in security-sensitive code.

I don't understand exactly. Is using `std::strlen` with a string literal not explicit behavior? Or is the maybe-not compile time an issue in this situation?

I've used `sizeof` liberally in places too, but it can be tricky since `MONERO_VERSION` is declared in another file. If the type is `const char*` instead of `const char[]`, then the wrong value is computed whereas `std::strlen` uses type checking and will compute the correct length either way (although with `const char*` it does so at runtime).

The likelihood of `sizeof` being used incorrectly here is low due to the way these constants are typically declared, so I'm not recommending against its usage either.

Reply:
I suppose `MONERO_VERSION` could also be a `nullptr` or incorrect pointer value, which breaks `std::strlen`. But I don't know how `sizeof` could work correctly in any situation where `MONERO_VERSION` refers to an incorrect pointer.